QgsFeaturePool: back to a write locker type

Djedouas opened the pull request:
## Description

The comment in the code clearly explains why we need to get a Write locker. I just run into this gotcha and set a Read locker...

Now back to a Write locker.

Fixes #61087

Djedouas committed:
fix(qgsfeaturepool): back to a write locker type

The comment clearly explains why we need to get a Write locker. I just
run into this gotcha and set a Read locker...

Now back to a Write locker.

Fixes #61087

nyalldawson commented on src/analysis/vector/geometry_checker/qgsfeaturepool.cpp:
```diff
@@ -59,7 +51,15 @@ bool QgsFeaturePool::getFeature( QgsFeatureId id, QgsFeature &feature )
+  // small print of the QCache docs. This is the hint that reality is different.
+  //
+  // https://bugreports.qt.io/browse/QTBUG-19794
+  QgsReadWriteLocker locker( mCacheLock, QgsReadWriteLocker::Write );
```
Evidently the giant warning comment wasn't sufficient already, so...

```suggestion
  /* DO NOT CHANGE */ QgsReadWriteLocker locker( mCacheLock, QgsReadWriteLocker::Write ); // DO NOT CHANGE!!!!!
```

nyalldawson committed:
Update src/analysis/vector/geometry_checker/qgsfeaturepool.cpp

nyalldawson commented:
As penance for not reading the giant warning comment you need to add a unit test preventing this breaking again :stuck_out_tongue_closed_eyes:

Djedouas committed:
fix(qgsfeaturepool): add multithreaded geometry checker test

This test tests the geometry checker method used in the geometry checker
plugin which launches the checks in parallel,
QgsGeometryChecker::execute().

We use a relatively big layer to be maximize chances of triggering
threading issues and be as close as possible to a real user case.

Related to #61087

Djedouas commented:
> As penance for not reading the giant warning comment you need to add a unit test preventing this breaking again 😝

Haha, right... :P  I added a commit with a test.

In fact, this multithreaded way of launching geometry checks was not tested at all.

I had to add a new test layer however, and took the one from the issue, because I couldn't reproduce the crash with the current test layers available in the geometry checker test data. I guess that the concurrent access does not occur due to the small amount of features.

I can reproduce the crash with this test every time (locally) with the new test layer and the ::Read lock type.

nyalldawson commented:
Hmm that's a fairly large reference file to add to the repo. Does zipping it help at all?

Djedouas commented:
Zipping helped a bit, ~25% smaller.

Thanks for the hint :)

Djedouas commented:
@nyalldawson Is it ok to merge?

troopa81 reviewed (approved):
This looks good to me!
